serveradmin/serverdb/views.py

```diff
@@ -1,181 +1,145 @@
+    commits = ChangeCommit.objects.all().order_by('-id')
     date_settings = {
-        'TIMEZONE': settings.TIME_ZONE,
-        'RETURN_AS_TIMEZONE_AWARE': True,
+        'TIMEZONE': settings.TIME_ZONE, 'RETURN_AS_TIMEZONE_AWARE': True
     }
 
-    try:
-        if hostname:
-            object_id = Server.objects.get(hostname=hostname).server_id
-        else:
-            object_id = request.GET.get('object_id')
-    except ObjectDoesNotExist:
-        messages.error(request, 'Server does not exist')
-        return TemplateResponse(request, 'serverdb/changes.html', {})
-
-    if t_from:
-        column_filter['change_on__gt'] = dateparser.parse(
-            t_from,
-            settings=date_settings,
-        )
-        context['from_understood'] = column_filter['change_on__gt']
-    if t_until:
-        column_filter['change_on__lt'] = dateparser.parse(
-            t_until,
-            settings=date_settings,
-        )
-        context['until_understood'] = column_filter['change_on__lt']
-    if object_id:
-        q_filter.append((
-            Q(changeupdate__server_id=object_id) |
-            Q(changedelete__server_id=object_id) |
-            Q(changeadd__server_id=object_id)
-        ))
-    if application:
-        q_filter.append((
-            Q(app__name=application) | Q(user__username=application)
-        ))
-
-    commits = ChangeCommit.objects.filter(
-        *q_filter,
-        **column_filter,
-    ).order_by('-change_on')
-    paginator = Paginator(commits, 20)
-
-    try:
-        page = paginator.page(request.GET.get('page', 1))
-    except (PageNotAnInteger, EmptyPage):
-        page = paginator.page(1)
-
-    context.update({
+    f_from = request.GET.get('from')
+    if f_from:
+        f_from = dateparser.parse(f_from, settings=date_settings)
+        commits = commits.filter(change_on__gt=f_from)
+
+    f_until = request.GET.get('until')
+    if f_until:
+        f_until = dateparser.parse(f_until, settings=date_settings)
+        commits = commits.filter(change_on__lt=f_until)
+
+    f_hostname = request.GET.get('hostname')
+    if f_hostname:
+        # Display all changes that have or had this hostname
+        # TODO: Create DB index for hostname to speed up this
+        object_ids = set(Change.objects.filter(
+            change_type__in=[Change.Type.CREATE, Change.Type.DELETE],
+            change_json__hostname=f_hostname).values_list(
+            'change_json__object_id', flat=True))
+        commits = commits.filter(change__object_id__in=object_ids)
+
+    f_object_id = request.GET.get('object_id')
+    if f_object_id:
+        commits = commits.filter(change__object_id=f_object_id)
+
+    f_user_or_app = request.GET.get('user_or_app')
+    if f_user_or_app:
+        commits = commits.filter(
+            Q(app__name=f_user_or_app) | Q(user__username=f_user_or_app))
+
+    commits = commits.select_related('app', 'user')
+    commits = commits.prefetch_related('change_set')
```

Comment:
You can change this line like so to fetch the relating `hostname` in the same query:
```
    server = Server.objects.filter(
        pk=OuterRef('object_id'),
    ).only('hostname')
    commits = commits.prefetch_related(Prefetch(
        'change_set',
        queryset=Change.objects.annotate(
            hostname=Subquery(server.values('hostname')),
        ),
    ))
```
However, I am not quite sure I understand your comment in the model inside the `except Server.DoesNotExist` clause. But if you could elaborate then we can maybe find a way to handle this case as well. Afterwards you can remove the whole `hostname` method of the `Change` model

Reply:
This wouldn't get us what we want as the hostname might not always be present when a object has been deleted. It might be able to adjust the code but I couldn't get it working within 30 minutes so I'll exclude this improvment from this release and have a look at it later.